### ISSUE TITLE
workbench.action.quickOpen (⌘P) switch between two files quicker

### ISSUE THREAD

**Issue by ScottRudiger** [microsoft/vscode#79680]
#### TL;DR: `⌘P, Enter` switches to previously viewed file (rather than `⌘P, ↓, Enter`)

Currently, one way to switch between two files quickly is to use **`⌘P, ↓, Enter`**. However, I would prefer the ability to save a keystroke here so that it is consistent with how I use [this Chrome extension](https://chrome.google.com/webstore/detail/quick-tabs/jnjfeinjfmenlddahdjdmgpbokiacbbb). It basically switches to the previously viewed tab with **`[shortcut], Enter`** and you can switch to other recent tabs very similarly to Quick Open.

So, I propose VSCode change the behavior of Quick Open to match and select the previously viewed file when opened or add an option such as:

```jsonc
// Controls whether the currently focused file or the previously focused file is selected by default when Quick Open is opened
//  - previous: The previously focused file is selected by default when Quick Open opens
//  - current: The currently focused file is selected by default when Quick Open opens
"workbench.quickOpen.defaultSelection": "previous",
```

### More reasoning to do this:

Currently **`⌘P, Enter`** opens the currently opened file. In essence, this is the same as just closing Quick Open and is the third way to do so (**`Escape`**, **`⇧ + Esc`**, and **`Enter`**). I would argue **`⌘P, Enter`** switching to the previously viewed file should be the default because of this.

**Happy to open a PR with some guidance on where to start for this feature. 💙**

### Other considerations:
I'm not sure how this feature should interact with:
-  `workbench.quickOpen.preserveInput`
- `search.quickOpen.includeHistory`
- `search.quickOpen.includeSymbols`

###### _(Sorry for the mac-only keyboard shortcuts; it's just what I'm familiar with with VSCode.)_

**Comment by RMacfarlane** [comment]
Hey @scottaddie, there's an existing command called `View: Open Previous Recently Used Editor` that you could add a keybinding to. (It's counterpart is `View: Open Next Recently Used Editor` to navigate the opposite direction in file open history). Would that accomplish what you want?

**Comment by ScottRudiger** [comment]
@RMacfarlane, the functionality of quickly switching between two tabs with `View: Open Next Recently Used Editor in Group` or `workbench.action.quickOpenNavigateNextInEditorPicker` is exactly what I'm looking for, except I'd like the ability to do so in the Quick Open feature with `⌘P, Enter` so I can use the same shortcut for multiple purposes:

1. `⌘P, Enter` - Switch to previously focused tab. Another `⌘P, Enter` would bring you back to the original tab.
2. `⌘P, ↓, ↓, ↓, Enter` - Switch to some other tab (normal operation of Quick Open)

Hope this makes sense!

**Comment by chrmarti** [comment]
You can already press `P` a second (or more) time while still holding `⌘` and then release all keys to switch to the currently selected file. Does that address your request?

**Comment by ScottRudiger** [comment]
@chrmarti I'm not sure I follow. For me, `⌘ + P` opens Quick Open. Then, if I keep holding down `⌘` and hit `P` again, it just closes Quick Open. (Neat! It seems this is the fourth way to close Quick Open.)

I don't believe this addresses my feature request. What I'd like to see is `⌘ + P` followed by `Enter` switch from the current tab to the previous tab. As of now, this just closes Quick Open and remains on the current tab.

In other words, I would like `⌘ + P, Enter` to behave exactly how `⌘ + P, ↓, Enter` behaves now. This would allow cycling between two tabs by repeating `⌘ + P, Enter`.

**Comment by chrmarti** [comment]
Not sure what the difference is between our setups, but subsequent `P` presses move the selection down for me. Do you have a custom keybinding for `⌘ + P`?

**Comment by chrmarti** [comment]
What is your `workbench.action.quickOpenNavigateNextInFilePicker` bound to? I have that on `⌘ + P` (the default) and that makes the selection move down when the file picker is already open.

When it works, it's the same number of keystrokes you are asking for because `Enter` is not needed in that case: `⌘ + P ... P`. Additionally you can keep pressing `P` to move down and adding `Shift` to it would move up on `P` (while `⌘` stays pressed, releasing it opens the file).

**Comment by ScottRudiger** [comment]
Thanks for the suggestion!

I will try setting it to the same key combination as Quick Open (`⌘ + E` in my case) and see if I can get used to the difference in behavior with the chrome extension. For some reason my  `workbench.action.quickOpenNavigateNextInFilePicker` was set to `^ + Tab`—perhaps something to do with pulling in Sublime keymaps at some point.

Totally understand if you decide this feature isn't worth prioritizing right now. Or, I'd be happy to submit a PR (given enough time).

**Comment by ScottRudiger** [comment]
Sorry, but I'm having some trouble making this work. I now have both `workbench.action.quickOpen` and `workbench.action.quickOpenNavigateNextInFilePicker` mapped to `⌘ + E`, but it doesn't seem to actually navigate to the next file:

![gC2a5XMvCL](https://user-images.githubusercontent.com/26824724/63921547-d9d17a00-c9f7-11e9-9a3a-8c53e4477699.gif)

Is this how yours looks (except both set to `⌘ + P`)?

**Comment by bpasero** [comment]
@ScottRudiger how is what you want not covered by this already?

![before](https://user-images.githubusercontent.com/900690/69010896-bbc74600-0964-11ea-90c7-8b54e8204974.gif)

**Comment by silvus** [comment]
"_workbench.action.openNextRecentlyUsedEditorInGroup_" do not have the same behavior: he select on key release, not on enter, and do not have filters capabilities.

On others editor (Sublime Text, Intellij, etc...) the current file is exclude from the recents files lists. It make sens has you can't navigate to an already open file.
For now, in VsCode with "_workbench.action.quickOpen_", the cursor is placed on the current file who doesn't seems to do anything on enter.

**Comment by bpasero** [comment]
@silvus you can do this:
* Cmd+P
* leave Cmd pressed
* type P until your file is selected
* release

Does that helpß

**Comment by silvus** [comment]
It's my current method but it's frustrating when you alternate with another editor to have to change keyboard shortcuts.
In the others I can do ctrl + p, here I have to use ctrl + p + p.

Since I have to use PHPStorm in parallel, the muscle memory is put to the test.
As for @ScottRudiger with its Chrome plugin, it's a consistency problem with other tools.

**Comment by bpasero** [comment]
> In the others I can do ctrl + p

How does that work exactly? `Ctrl+P` once to open quick open and then `Ctrl+P` again to open the selected entry? Assume I have never used any other editor than VSCode in my life.

**Comment by silvus** [comment]
No, to be precise I use ctrl+P, then validate with Enter.
The number of keys pressed is the same (but they are not the same :wink: ).

The first item in the list, the one selected, is the file opened previously.
So if I use ctrl + p, then Enter 2 times, I'm back on the starting file.

I can formulate the problem differently: I do not understand the interest of putting the cursor on the current file, because by default, ctrl + p then validate with Enter does nothing.
Is there a reason for this behavior? Or, why have the current file in the window "workbench.action.quickOpen"?

The quick fix that would make sense here is maybe just not to position the cursor on the current file on opening, but on the one previously in the history? Or filter the current file from the list?

**Comment by bpasero** [comment]
@silvus so this sounds a lot like https://github.com/microsoft/vscode/issues/6923 where we added a specific command for that purpose that you can bind to: `workbench.action.quickOpenPreviousEditor`

**Comment by silvus** [comment]
That's it ! Awesome! Thank you !
I do not understand why I did not find this command before...

```json
{
	"key": "ctrl+p",
	"command": "workbench.action.quickOpenPreviousEditor"
},
{
	"key": "ctrl+p",
	"command": "-workbench.action.quickOpen"
},
```

**Comment by bpasero** [comment]
Closing, the answer is https://github.com/microsoft/vscode/issues/79680#issuecomment-564564115
